File: prediction-service/app/services/api_football.py
"""
API-Football integration service for Python prediction engine
"""

import httpx
import asyncio
from typing import List, Dict, Optional, Any
from loguru import logger

from app.core.config import settings


class ApiFootballService:
    def __init__(self):
        self.base_url = f"https://{settings.API_FOOTBALL_HOST}"
        self.headers = {
            "x-rapidapi-key": settings.API_FOOTBALL_KEY,
            "x-rapidapi-host": settings.API_FOOTBALL_HOST,
        }
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            headers=self.headers,
            timeout=10.0,
            limits=httpx.Limits(max_connections=20, max_keepalive_connections=10),
        )

    async def _make_request(self, endpoint: str, params: Dict[str, Any] = None) -> Dict:
        """Make API request with retry logic"""
        max_retries = 3
        for attempt in range(max_retries):
            try:
                response = await self.client.get(endpoint, params=params)
                response.raise_for_status()
                data = response.json()

                if data.get("errors"):
                    logger.warning(f"API errors: {data['errors']}")

                return data.get("response", [])
            except httpx.HTTPStatusError as e:
                if e.response.status_code == 429:
                    wait_time = 2 ** attempt
                    logger.warning(f"Rate limited. Waiting {wait_time}s...")
                    await asyncio.sleep(wait_time)
                else:
                    raise
            except Exception as e:
                if attempt == max_retries - 1:
                    logger.error(f"Failed after {max_retries} attempts: {e}")
                    raise
                await asyncio.sleep(1)
        return []

    async def get_fixture(self, fixture_id: int) -> Optional[Dict]:
        """Get fixture by ID"""
        data = await self._make_request("/fixtures", {"id": fixture_id})
        return data[0] if data else None

    async def get_team_form(self, team_id: int, last: int = 5, season: Optional[int] = None) -> List[Dict]:
        """Get team's recent form"""
        params = {"team": team_id, "last": last}
        if season:
            params["season"] = season

        fixtures = await self._make_request("/fixtures", params)

        form = []
        for fixture in fixtures:
            home_team = fixture["teams"]["home"]["id"]
            home_goals = fixture["goals"]["home"] or 0
            away_goals = fixture["goals"]["away"] or 0

            is_home = home_team == team_id
            team_goals = home_goals if is_home else away_goals
            opp_goals = away_goals if is_home else home_goals

            if team_goals > opp_goals:
                result = "W"
            elif team_goals == opp_goals:
                result = "D"
            else:
                result = "L"

            form.append({
                "fixture_id": fixture["fixture"]["id"],
                "result": result,
                "goals_for": team_goals,
                "goals_against": opp_goals,
                "venue": "home" if is_home else "away",
                "opponent": fixture["teams"]["away" if is_home else "home"]["name"],
                "date": fixture["fixture"]["date"],
            })

        return form

    async def get_head_to_head(self, team1: int, team2: int, last: int = 10) -> List[Dict]:
        """Get head-to-head record"""
        data = await self._make_request(
            "/fixtures/headtohead",
            {"h2h": f"{team1}-{team2}", "last": last}
        )

        h2h = []
        for fixture in data:
            h2h.append({
                "fixture_id": fixture["fixture"]["id"],
                "home_team_id": fixture["teams"]["home"]["id"],
                "away_team_id": fixture["teams"]["away"]["id"],
                "home_goals": fixture["goals"]["home"] or 0,
                "away_goals": fixture["goals"]["away"] or 0,
                "date": fixture["fixture"]["date"],
                "league": fixture["league"]["name"],
            })

        return h2h

    async def get_injuries(self, team_id: int) -> List[Dict]:
        """Get team injuries"""
        data = await self._make_request("/injuries", {"team": team_id})

        injuries = []
        for injury in data:
            injuries.append({
                "player_id": injury["player"]["id"],
                "player_name": injury["player"]["name"],
                "type": injury["player"]["type"],
                "reason": injury["player"].get("reason", ""),
            })

        return injuries

    async def get_standings(self, league_id: int, season: int) -> List[Dict]:
        """Get league standings"""
        data = await self._make_request("/standings", {"league": league_id, "season": season})

        if data and len(data) > 0:
            return data[0]["league"]["standings"][0] if data[0]["league"]["standings"] else []
        return []

    async def get_fixtures_by_date_range(
        self, 
        date_range: str, 
        leagues: Optional[List[int]] = None
    ) -> List[Dict]:
        """Get fixtures for date range"""
        from datetime import datetime, timedelta

        today = datetime.now()

        if date_range == "today":
            date = today.strftime("%Y-%m-%d")
            return await self._make_request("/fixtures", {"date": date, "league": leagues or "", "season": today.year})
        elif date_range == "tomorrow":
            date = (today + timedelta(days=1)).strftime("%Y-%m-%d")
            return await self._make_request("/fixtures", {"date": date, "league": leagues or "", "season": today.year})
        elif date_range == "week":
            fixtures = []
            for i in range(7):
                date = (today + timedelta(days=i)).strftime("%Y-%m-%d")
                day_fixtures = await self._make_request("/fixtures", {"date": date, "league": leagues or "", "season": today.year})
                fixtures.extend(day_fixtures)
            return fixtures

        return []

    async def close(self):
        await self.client.aclose()

File: backend/src/services/api-football.ts
"use server";

import axios from "axios";
import axiosRetry from "axios-retry";

const RAPIDAPI_KEY = process.env.RAPIDAPI_KEY;
const RAPIDAPI_HOST = process.env.RAPIDAPI_HOST || "api-football-v1.p.rapidapi.com";
const RAPIDAPI_FOOTBALL_ENDPOINT = process.env.RAPIDAPI_FOOTBALL_ENDPOINT || "https://api-football-v1.p.rapidapi.com";

const apiFootballClient = axios.create({
  baseURL: RAPIDAPI_FOOTBALL_ENDPOINT,
  headers: {
    "x-rapidapi-key": RAPIDAPI_KEY,
    "x-rapidapi-host": RAPIDAPI_HOST,
  },
  timeout: 10000,
});

// Retry logic for API failures
axiosRetry(apiFootballClient, {
  retries: 3,
  retryDelay: axiosRetry.exponentialDelay,
  retryCondition: (error) => {
    return axiosRetry.isNetworkOrIdempotentRequestError(error) || 
           (error.response?.status ?? 0) >= 500;
  },
});

export interface ApiFootballFixture {
  fixture: {
    id: number;
    referee: string | null;
    timezone: string;
    date: string;
    timestamp: number;
    status: {
      short: string;
      elapsed: number | null;
    };
    venue: { name: string | null; city: string | null };
  };
  league: {
    id: number;
    name: string;
    country: string;
    season: number;
    logo: string;
    flag: string | null;
  };
  teams: {
    home: { id: number; name: string; logo: string };
    away: { id: number; name: string; logo: string };
  };
  goals: { home: number | null; away: number | null };
  score: {
    halftime: { home: number | null; away: number | null };
    fulltime: { home: number | null; away: number | null };
    extratime: { home: number | null; away: number | null };
    penalty: { home: number | null; away: number | null };
  };
}

export interface ApiFootballStanding {
  rank: number;
  team: { id: number; name: string; logo: string };
  points: number;
  goalsDiff: number;
  group: string;
  form: string;
  status: string;
  all: {
    played: number;
    win: number;
    draw: number;
    lose: number;
    goals: { for: number; against: number };
  };
  home: {
    played: number;
    win: number;
    draw: number;
    lose: number;
    goals: { for: number; against: number };
  };
  away: {
    played: number;
    win: number;
    draw: number;
    lose: number;
    goals: { for: number; against: number };
  };
}

export interface ApiFootballH2H {
  fixture: {
    id: number;
    date: string;
    timestamp: number;
  };
  league: { name: string };
  teams: {
    home: { id: number; name: string; logo: string; winner: boolean | null };
    away: { id: number; name: string; logo: string; winner: boolean | null };
  };
  goals: { home: number | null; away: number | null };
}

export interface ApiFootballInjury {
  player: { id: number; name: string; photo: string };
  team: { id: number; name: string; logo: string };
  fixture: { id: number; date: string };
  league: { id: number };
  type: string;
  reason: string;
}

export interface ApiFootballStatistic {
  team: { id: number; name: string; logo: string };
  statistics: Array<{ type: string; value: string | number | null }>;
}

/**
 * Fetch fixtures by date range
 */
export async function getFixtures(
  date: string,
  league?: number,
  season?: number,
  status?: string
): Promise<ApiFootballFixture[]> {
  const params: Record<string, string | number> = { date };
  if (league) params.league = league;
  if (season) params.season = season;
  if (status) params.status = status;

  const response = await apiFootballClient.get("/fixtures", { params });
  return response.data.response || [];
}

/**
 * Fetch fixture by ID with all details
 */
export async function getFixtureById(fixtureId: number): Promise<ApiFootballFixture | null> {
  const response = await apiFootballClient.get("/fixtures", {
    params: { id: fixtureId },
  });
  const data = response.data.response;
  return data && data.length > 0 ? data[0] : null;
}

/**
 * Fetch league standings
 */
export async function getStandings(
  league: number,
  season: number
): Promise<ApiFootballStanding[]> {
  const response = await apiFootballClient.get("/standings", {
    params: { league, season },
  });
  const data = response.data.response;
  return data && data.length > 0 ? data[0].league.standings[0] : [];
}

/**
 * Fetch head-to-head between two teams
 */
export async function getHeadToHead(
  team1Id: number,
  team2Id: number,
  last: number = 10
): Promise<ApiFootballH2H[]> {
  const h2h = `${team1Id}-${team2Id}`;
  const response = await apiFootballClient.get("/fixtures/headtohead", {
    params: { h2h, last },
  });
  return response.data.response || [];
}

/**
 * Fetch team injuries
 */
export async function getInjuries(
  team: number,
  fixture?: number
): Promise<ApiFootballInjury[]> {
  const params: Record<string, number> = { team };
  if (fixture) params.fixture = fixture;

  const response = await apiFootballClient.get("/injuries", { params });
  return response.data.response || [];
}

/**
 * Fetch fixture statistics
 */
export async function getFixtureStatistics(
  fixture: number,
  team?: number
): Promise<ApiFootballStatistic[]> {
  const params: Record<string, number> = { fixture };
  if (team) params.team = team;

  const response = await apiFootballClient.get("/fixtures/statistics", { params });
  return response.data.response || [];
}

/**
 * Fetch team form (last N fixtures)
 */
export async function getTeamForm(
  teamId: number,
  last: number = 5,
  season?: number
): Promise<ApiFootballFixture[]> {
  const params: Record<string, number> = { team: teamId, last };
  if (season) params.season = season;

  const response = await apiFootballClient.get("/fixtures", { params });
  return response.data.response || [];
}

/**
 * Fetch all available leagues
 */
export async function getLeagues(
  season?: number,
  country?: string
): Promise<any[]> {
  const params: Record<string, string | number> = {};
  if (season) params.season = season;
  if (country) params.country = country;

  const response = await apiFootballClient.get("/leagues", { params });
  return response.data.response || [];
}

/**
 * Fetch live fixtures
 */
export async function getLiveFixtures(): Promise<ApiFootballFixture[]> {
  const response = await apiFootballClient.get("/fixtures", {
    params: { live: "all" },
  });
  return response.data.response || [];
}
